README.md
# StatsD Instrumental Backend


## Overview

Instrumental is a [application monitoring platform](https://instrumentalapp.com) built for developers who want a better understanding of their production software. Powerful tools, like the [Instrumental Query Language](https://instrumentalapp.com/docs/query-language), combined with an exploration-focused interface allow you to get real answers to complex questions, in real-time.

This is a pluggable StatsD backend for sending metrics to Instrumental.

## Requirements

* [StatsD][statsd] versions >= 0.3.0.
* An [Instrumental](https://instrumentalapp.com) account.

## Installation

    $ cd /path/to/statsd
    $ npm install statsd-instrumental-backend

## Configuration

See our [example config file](exampleConfig.js) for a complete StatsD configuration.

Otherwise, add the following basic configuration information to your
StatsD config file.

```js
{
  instrumental: {
    key: "[project token]",       // REQUIRED
    secure: true,                 // OPTIONAL (boolean), whether or not to use secure protocol to connect to Instrumental, default true
    verifyCert: true,             // OPTIONAL (boolean), should we attempt to verify the server certificate before allowing communication, default true
    timeout: 10000,               // OPTIONAL (integer), number of milliseconds to wait for establishing a connection to Instrumental before giving up, default 10s
    recordCounterRates: true,     // OPTIONAL (boolean) whether or not to send ".rate" metrics with counters, default true
    metricPrefix: "",             // OPTIONAL (string) this will be prepended (with a dot) to ALL of your metrics
    metricFiltersExclude: [],     // OPTIONAL (array of regexes) any metrics matching these filters will be dropped. e.g, [/\.rate$/, /\.[^\.]*_99$/] would filter out counter rates and extra metrics for stats like upper_99, mean_99, etc.
    metricFiltersInclude: []      // OPTIONAL (array of regexes) any metrics _NOT_ matching at least one of these filters will be dropped
  }
}
```

### Filtering Metrics

Metrics can be filtered with either the `metricFiltersExclude` or `metricFiltersInclude` options. Note that for a metric to be sent, it must match at least one include *and* no excludes.  An empty or missing metricFiltersInclude option will allow all metrics (assuming the exclusion filter does not exclude any).

## Enabling

This is already done if you are using our [example configuration](exampleConfig.js).

Otherwise, add `statsd-instrumental-backend` backend to the list of StatsD
backends in the StatsD configuration file:

```js
{
  backends: ["statsd-instrumental-backend"]
}
```

Start/restart the statsd daemon and your StatsD metrics should now be
pushed to your Instrumental account.

## NPM Dependencies

None

## Development

- [Instrumental Backend](https://github.com/expectedbehavior/statsd-instrumental-backend)

Contributing:

* Fork the project
* Make your feature addition or bug fix
* Commit. Do not mess with package.json, version, or history.
* Send a pull request. Bonus points for topic branches.

# Releasing

1. Sign up for npmjs.com
2. Pull latest master
3. Merge feature branch(es) into master
4. `./script/test`
5. Increment version in `package.json`
6. Update [CHANGELOG.md](CHANGELOG.md)
7. Commit "Release version vX.Y.Z"
8. Push to GitHub
9. Tag version: `git tag 'vX.Y.Z' && git push --tags`
10. `npm publish`
11. Verify update on https://www.npmjs.com/package/statsd-instrumental-backend

[statsd]: https://github.com/etsy/statsd


CHANGELOG.md
### 0.13.1 [November 28th, 2018]
* Reduce logging with production debug settings

### 0.13.0 [May 10th, 2018]
* Add new certificates that will work after the current intermediate certs expire later this year
* Add a configuration option to enable custom certificates
* Add the ability to use included node certs, and make it the default (with included certs as a fallback)

### 0.12.4 [March 23rd, 2017]
* Revert to string.match from regex.test.  String.match is available in more environments.

### 0.12.3 [March 21st, 2017]
* Remove some logging that snuck in to production

### 0.12.2 [March 21st, 2017]
* Make metric whitelisting functionality (`metricFiltersInclude`) more efficient in the no-op case

### 0.12.1 [March 21st, 2017]
* Introduce metric whitelist/blacklist via `metricFiltersInclude` and `metricFiltersExclude` options

### 0.11.4 [December 9th, 2016]
* Fix a bug that caused `metricPrefix` to not always be applied.

### 0.11.3 [December 8th, 2016]
* Add `metricPrefix` option, which will be added to all of your metrics (with a "." to seperate them)

### 0.11.2 [March 28th, 2016]
* Add recordCounterRates option for users who wish to ignore .rate metrics on counters

### 0.11.1 [June 24th, 2015]
* Send reporting host information in transport greeting
* Better buffered data handling

### 0.11.0 [June 2nd, 2015]
* Add support for encrypted transport (TLS)
* Transition ownership to Expected Behavior


test/ssl_test.js
var TestHelper = require("./test_helper");
var timekeeper = require('timekeeper');
var path = require("path");
var util = require("util");
var fs   = require("fs");

var test = TestHelper.test;
var setup = TestHelper.setup;
var sendMetric = TestHelper.sendMetric;
var checkForMetric = TestHelper.checkForMetric;

test('specifying a valid and working cert bundle works', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  var certs = [];
  var certDir = path.join(__dirname, "..", "certs/");
  fs.readdirSync(certDir).forEach(function(filename) {
    if (filename.match(/ca\.pem$/) && !filename.match(/20/)) {
      certs.push(certDir + filename);
    }
  });

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  TestHelper.config.instrumental.caCertFiles = certs;
  now = Math.round(new Date().getTime() / 1000);
  var metricName = "test.metric"+Math.random();
  sendMetric(metricName, oldTime);

  checkForMetric(metricName, {
    found: function(){
      t.pass();
      t.end();
    },
    timeout: function(){
      t.fail();
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});

test('specifying a valid but not working cert bundle retries', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  var certs = [];
  var certDir = path.join(__dirname, "..", "certs/");
  fs.readdirSync(certDir).forEach(function(filename) {
    if (filename.match(/ca\.pem$/) && filename.match(/20/)) {
      certs.push(certDir + filename);
    }
  });

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  TestHelper.config.instrumental.caCertFiles = certs;
  now = Math.round(new Date().getTime() / 1000);
  var metricName = "test.metric"+Math.random();
  sendMetric(metricName, oldTime);
  sendMetric(metricName, oldTime, {skipInit: true});
  var checkConnectionErrors = function() {
    var connection_errors =
      TestHelper.log.filter(function(entry){return entry.match(/Client error:/)});
    t.equal(connection_errors.length, 2, "expected 2 connection attemps, 1 retry");
    t.end();
  };
  setTimeout(checkConnectionErrors, 1000);
});

test('specifying an invalid cert bundle errors', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  TestHelper.config.instrumental.caCertFiles = ["non_existent_file"];
  var metricName = "test.metric"+Math.random();
  t.throws(function(){
    sendMetric(metricName, oldTime);
  }, /no such file/, "expected error finding non_existent_file");
  t.end();
});

test('node default certs are used by default', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  var metricName = "test.metric"+Math.random();
  sendMetric(metricName, oldTime);

  checkForMetric(metricName, {
    found: function(){
      var cert_log_messages =
        TestHelper.log.filter(function(entry){return entry.match(/Using certs/i)});
      t.deepEqual(cert_log_messages, ["Using certs: node default"], "expected to use node default certs");
      t.pass();
      t.end();
    },
    timeout: function(){
      t.fail();
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});

test('before cert expiration, fallback to newest cert bundle and stick if node default certs fail', function(t) {
  setup(t);

  realCurrentTime = Math.round(new Date().getTime() / 1000);

  var time = new Date(Date.parse("2018-01-01"));
  timekeeper.travel(time); // Travel to that date.

  var metricName = "test.metric"+Math.random();

  TestHelper.config.instrumental.tlsVariationTimeout = 500;

  var timeBetweenSends = 2000;
  var actions = [];

  // node default certs
  actions.push(function(){
    TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
    sendMetric(metricName, realCurrentTime); // using node default certs
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // using node default certs
  });

  // New cert bundle
  actions.push(function(){
    TestHelper.config.instrumental.port = 8000;
    sendMetric(metricName, realCurrentTime); // node default certs fail
  });
  actions.push(function(){
    TestHelper.config.instrumental.port = 8001;
    sendMetric(metricName, realCurrentTime); // bundle
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // bundle
  });

  // Back to node default
  actions.push(function(){
    TestHelper.config.instrumental.port = 8000;
    sendMetric(metricName, realCurrentTime); // bundle fails
  });
  actions.push(function(){
    TestHelper.config.instrumental.port = 8001;
    sendMetric(metricName, realCurrentTime); // node default
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // node default
  });

  actions.forEach(function(action, index){
    setTimeout(action, index*timeBetweenSends);
  });

  checkForMetric(metricName, {
    expectedSum: 6, // failures don't get sent
    found: function(){
      var cert_log_messages =
        TestHelper.log.filter(function(entry){return entry.match(/\bcert/i)});
      var expected_messages = [
        "Adding node default ssl cert option",
        'Found valid cert bundle: equifax.2018-08-19',
        'Found valid cert bundle: geotrust.2018-08-19',
        'Found valid cert bundle: rapidssl.2018-08-19',
        'Found valid cert bundle: digicert_intermediate',
        'Found valid cert bundle: digicert_root',
        'Found valid cert bundle: rapidssl',
        "Attempting new cert config: node default",
        "Using certs: node default", // 1
        "Using certs: node default", // 2
        "Using certs: node default", // fail
        "Attempting new cert config: bundles equifax.2018-08-19 geotrust.2018-08-19 rapidssl.2018-08-19 digicert_intermediate digicert_root rapidssl",

        // 3
        "Using certs: bundles equifax.2018-08-19 geotrust.2018-08-19 rapidssl.2018-08-19 digicert_intermediate digicert_root rapidssl",

        // 4
        "Using certs: bundles equifax.2018-08-19 geotrust.2018-08-19 rapidssl.2018-08-19 digicert_intermediate digicert_root rapidssl",

        // fail
        "Using certs: bundles equifax.2018-08-19 geotrust.2018-08-19 rapidssl.2018-08-19 digicert_intermediate digicert_root rapidssl",
        "Attempting new cert config: node default",
        "Using certs: node default", // 5
        "Using certs: node default"  // 6
      ];
      t.deepEqual(cert_log_messages, expected_messages, "expected to use node default certs");
      t.pass();
      t.end();
    },
    timeout: function(){
      var cert_log_messages =
        TestHelper.log.filter(function(entry){return entry.match(/Using certs/i)});
      t.fail(JSON.stringify(cert_log_messages) + "\n\n");
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});

test('after cert expiration, fallback to newest cert bundle and stick if node default certs fail', function(t) {
  setup(t);

  realCurrentTime = Math.round(new Date().getTime() / 1000);

  var time = new Date(Date.parse("2019-01-01"));
  timekeeper.travel(time); // Travel to that date.


  var metricName = "test.metric"+Math.random();

  TestHelper.config.instrumental.tlsVariationTimeout = 500;

  var timeBetweenSends = 2000;
  var actions = [];

  // node default certs
  actions.push(function(){
    TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
    sendMetric(metricName, realCurrentTime); // using node default certs
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // using node default certs
  });

  // New cert bundle
  actions.push(function(){
    TestHelper.config.instrumental.port = 8000;
    sendMetric(metricName, realCurrentTime); // node default certs fail
  });
  actions.push(function(){
    TestHelper.config.instrumental.port = 8001;
    sendMetric(metricName, realCurrentTime); // bundle
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // bundle
  });

  // Back to node default
  actions.push(function(){
    TestHelper.config.instrumental.port = 8000;
    sendMetric(metricName, realCurrentTime); // bundle fails
  });
  actions.push(function(){
    TestHelper.config.instrumental.port = 8001;
    sendMetric(metricName, realCurrentTime); // node default
  });
  actions.push(function(){
    sendMetric(metricName, realCurrentTime, {skipInit: true}); // node default
  });

  actions.forEach(function(action, index){
    setTimeout(action, index*timeBetweenSends);
  });

  checkForMetric(metricName, {
    expectedSum: 6, // failures don't get sent
    found: function(){
      var cert_log_messages =
        TestHelper.log.filter(function(entry){return entry.match(/\bcert/i)});
      var expected_messages = [
        "Adding node default ssl cert option",
        'Ignoring expired cert bundle: equifax.2018-08-19',
        'Ignoring expired cert bundle: geotrust.2018-08-19',
        'Ignoring expired cert bundle: rapidssl.2018-08-19',
        'Ignoring expired cert bundle: equifax.2018-08-19',
        'Ignoring expired cert bundle: geotrust.2018-08-19',
        'Ignoring expired cert bundle: rapidssl.2018-08-19',
        'Found valid cert bundle: digicert_intermediate',
        'Found valid cert bundle: digicert_root',
        'Found valid cert bundle: rapidssl',
        "Attempting new cert config: node default",
        "Using certs: node default", // 1
        "Using certs: node default", // 2
        "Using certs: node default", // fail
        "Attempting new cert config: bundles digicert_intermediate digicert_root rapidssl",

        // 3
        "Using certs: bundles digicert_intermediate digicert_root rapidssl",

        // 4
        "Using certs: bundles digicert_intermediate digicert_root rapidssl",

        // fail
        "Using certs: bundles digicert_intermediate digicert_root rapidssl",
        "Attempting new cert config: node default",
        "Using certs: node default", // 5
        "Using certs: node default"  // 6
      ];
      t.deepEqual(cert_log_messages, expected_messages, "expected to use node default certs");
      t.pass();
      t.end();
    },
    timeout: function(){
      var cert_log_messages =
        TestHelper.log.filter(function(entry){return entry.match(/Using certs/i)});
      t.fail(JSON.stringify(cert_log_messages) + "\n\n");
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});

test('old agent works with new elb', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  var time = new Date(Date.parse("2019-01-01"));
  timekeeper.travel(time); // Travel to that date.

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  var metricName = "test.metric"+Math.random();
  sendMetric(metricName, oldTime);

  checkForMetric(metricName, {
    found: function(){
      t.pass();
      t.end();
    },
    timeout: function(){
      t.fail();
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});

// Now that we'e past the switch date the old ELB doesn't exist. If we need to
// do this switch again in the future we can use this test to verify behavior.
// test('old agent connects to old elb', function(t) {
//   setup(t);
//
//   realCurrentTime = Math.round(new Date().getTime() / 1000);
//
//   var time = new Date(Date.parse("2018-01-01"));
//   timekeeper.travel(time); // Travel to that date.
//
//   TestHelper.config.instrumental.host = "collector.instrumentalapp.com";
//   now = Math.round(new Date().getTime() / 1000);
//   var metricName = "test.metric"+Math.random();
//   sendMetric(metricName, realCurrentTime);
//
//   checkForMetric(metricName, {
//     found: function(){
//       t.pass();
//       t.end();
//     },
//     timeout: function(){
//       t.fail();
//       t.end();
//     },
//     error: function(){
//       t.fail();
//       t.end();
//     },
//   });
// });

// Now that we'e past the switch date the old ELB doesn't exist. If we need to
// do this switch again in the future we can use this test to verify behavior.
// test('future agent correctly expires cert and errors with old elb', function(t) {
//   setup(t);
//
//   oldTime = Math.round(new Date().getTime() / 1000);
//
//   var time = new Date(Date.parse("2019-01-01"));
//   timekeeper.travel(time); // Travel to that date.
//
//   TestHelper.config.instrumental.disallowNodeDefaultCerts = true;
//   TestHelper.config.instrumental.host = "old-elb.instrumentalapp.com";
//   now = Math.round(new Date().getTime() / 1000);
//   var metricName = "test.metric"+Math.random();
//   sendMetric(metricName, oldTime);
//
//   checkForMetric(metricName, {
//     found: function(){
//       t.fail("Future agent shouldn't work with the old ELB, logs: " + TestHelper.log);
//       t.end();
//     },
//     timeout: function(){
//       var cert_log_messages =
//         TestHelper.log.filter(function(entry){
//           return entry.match(/\bcert/i) &&
//             !entry.match(/Error: unable to get/i) &&
//             !entry.match(/Error: CERT_UNTRUSTED/i) &&
//             !entry.match(/Error: certificate not trusted/i);
//         });
//       var expected_messages = [
//         "Skipping node default certificates",
//         'Found valid cert bundle: digicert_intermediate',
//         'Found valid cert bundle: digicert_root',
//         'Found valid cert bundle: rapidssl',
//         "Attempting new cert config: bundles digicert_intermediate digicert_root rapidssl",
//         "Using certs: bundles digicert_intermediate digicert_root rapidssl",
//       ];
//       t.deepEqual(cert_log_messages, expected_messages, "expected to use node default certs");
//       t.pass();
//       t.end();
//     },
//     error: function(){
//       t.fail();
//       t.end();
//     },
//   });
// });

test('future agent correctly expires cert and works with new elb', function(t) {
  setup(t);

  oldTime = Math.round(new Date().getTime() / 1000);

  var time = new Date(Date.parse("2019-01-01"));
  timekeeper.travel(time); // Travel to that date.

  TestHelper.config.instrumental.host = "smoke-collector.instrumentalapp.com";
  now = Math.round(new Date().getTime() / 1000);
  var metricName = "test.metric"+Math.random();
  sendMetric(metricName, oldTime);

  checkForMetric(metricName, {
    found: function(){
      t.pass();
      t.end();
    },
    timeout: function(){
      t.fail();
      t.end();
    },
    error: function(){
      t.fail();
      t.end();
    },
  });
});


lib/instrumental.js
/*
 * Flushes stats to Instrumental (https://instrumentalapp.com).
 *
 * To enable this backend, include 'statsd-instrumental-backend' in the
 * backends configuration array:
 *
 *   backends: ['statsd-instrumental-backend']
 *
 * The backend will read the configuration options from the following
 * 'instrumental' hash defined in the main statsd config file:
 *
 *  instrumental : {
 *    key     : API Token for your application,
 *    host    : Instrumental server address (default: "collector.instrumentalapp.com")
 *    port    : Instrumental server port (default: 8001)
 *    timeout : Connection and read timeout (default: 10000 [10s])
 *  }
 */

// TODO: Log some internal stats (error rate, reconnect, metrics sent)

var net  = require("net"),
    util = require("util"),
    os = require("os"),
    dns = require("dns"),
    package = require('../package.json'),
    tls  = require("tls"),
    fs   = require("fs"),
    path = require("path"),
    buffering_aggregation = require("./instrumental/buffering_aggregation"),
    p    = require("./instrumental/protocol");

var debug;
var error = false;
var hostname = os.hostname();
var caChainPath = path.join("..", "certs");
var knownCerts  = [
  "equifax.2018-08-19", "geotrust.2018-08-19", "rapidssl.2018-08-19",
  "digicert_intermediate", "digicert_root", "rapidssl"
];
var exports = module.exports = {}; // for testing
var key, host, port, timeout, flushInterval, secure, verifyCert, caChain,
  recordCounterRates, metricPrefix, metricFiltersInclude, metricFiltersExclude,
  tlsOptionsVariations, tlsVariationTimeout;
var formatMetricName = function(metric){ return metric; }
var keepMetric = function(metric){ return true; }

var instrumentalStats = {};

var log = function() {
  console.log.apply(null, arguments);
  if (debug) console.warn.apply(null, arguments);
}

function build_payload(metrics, time_stamp) {
  var payload = [];
  // Iterators: key and value.
  var k, v, name, vps;

  // increment item.count 200 121820381
  for(k in metrics.counters) {
    name = formatMetricName(k);
    v = metrics.counters[k];

    // Skip any counters that would not change the value
    if(v !== 0) {
      payload.push(p.inc(name, v, time_stamp));
    }

    if(recordCounterRates) {
      vps = metrics.counter_rates[k];
      payload.push(p.gauge_abs([name, "rate"].join("."), vps, time_stamp));
    }
  }

  // gauge item.time 7292 121820381
  for(k in metrics.timer_data){
    name = formatMetricName(k);
    for (var timer_data_key in metrics.timer_data[k]) {
      if (typeof(metrics.timer_data[k][timer_data_key]) === 'number') {
        payload.push(p.gauge_abs([name, timer_data_key].join("."), metrics.timer_data[k][timer_data_key], time_stamp));
      } else {
        for (var timer_data_sub_key in metrics.timer_data[k][timer_data_key]) {
          v = metrics.timer_data[k][timer_data_key][timer_data_sub_key];
          if(timer_data_sub_key.indexOf("bin") == 0){
            // This is histogram data, just increment the known server values
            payload.push(p.inc([name, timer_data_sub_key].join("."), v, time_stamp));
          } else {
            payload.push(p.gauge_abs([name, timer_data_sub_key].join("."), v, time_stamp));
          }

        }
      }
    }
  }

  // gauge item.gauge 7292 121820381
  for(k in metrics.gauges) {
    name = formatMetricName(k);
    payload.push(p.gauge(name, metrics.gauges[k], time_stamp));
  }

  // sets are strange and may not work correctly, but also, difficult to test.
  for(k in metrics.sets) {
    name = formatMetricName(k);
    payload.push(p.gauge_abs(name, metrics.sets[k].length, time_stamp));
  }

  payload = payload.filter(keepMetric);

  // By this point in time we should have an array of commands to send to the
  // instrumental server. We'll convert that array into a string blob in the
  // appropriate protocol.
  return payload;
}
exports.build_payload = build_payload;

function instrumental_connection(host, port, tlsOptionsVariation, onConnectCb){
  var connection;
  if (secure){
    if (debug) log("Using certs: " + tlsOptionsVariation.name);
    connection = tls.connect(port, host, tlsOptionsVariation.options, function(){
      if(verifyCert && !connection.authorized){
        throw new Exception("Instrumental connection cannot be encrypted, certificate not secured");
      }
      onConnectCb(connection);
    });
  } else {
    connection = net.createConnection(port, host);
    connection.on("connect", function(){
      onConnectCb(connection);
    });
  }
  // All communication with the server is done in ASCII.
  connection.setEncoding("ascii");
  return connection;
}

function shouldTryTlsOptionsVariation(variation) {
  var lastAttemptedSucceeded =
    variation.lastSuccessAt > variation.lastAttemptedAt;
  var lastAttemptedHasntFailedYet =
    (variation.lastAttemptedAt - variation.lastSuccessAt) < tlsVariationTimeout;
  return lastAttemptedSucceeded || lastAttemptedHasntFailedYet;
}

function selectTlsOptionsVariation() {
  // Find the first variation that looks successful, if one does
  var successfulVariation = tlsOptionsVariations.filter(function(variation){
    if ( shouldTryTlsOptionsVariation(variation) ) {
      variation.lastAttemptedAt = new Date();
      if (debug) log("Sticking with tls variation: ", variation);
      return true;
    }
  });
  if (successfulVariation[0]) { return successfulVariation[0]; };

  // Sort oldest attempt first, that'll be what we try next. Oldest so that we
  // eventually rotate through them all instead of flipping between two.
  tlsOptionsVariations.sort(function(a, b){
    return a.lastAttemptedAt - b.lastAttemptedAt;
  });

  var nextVariation = tlsOptionsVariations[0];
  log("Attempting new cert config: " + nextVariation.name);
  if (debug) log({tlsOptionsVariations: tlsOptionsVariations, tlsVariationTimeout: tlsVariationTimeout});

  // This is the first time we've tried this in a while so assume success
  // for the time being.
  nextVariation.lastAttemptedAt = new Date();
  nextVariation.lastSuccessAt = new Date();
  return nextVariation;
}

// Push data to instrumental
function instrumental_send(payload) {
  var state = "cold";
  var tlsOptionsVariation = selectTlsOptionsVariation();

  var client = instrumental_connection(host, port, tlsOptionsVariation, function(_client){
    state = "connected";

    var cleanString = function(value) {
      return String(value).replace(/\s+/g, "_");
    }

    // Write the authentication header
    _client.write(
      "hello version node/statsd-instrumental-backend/" + cleanString(package.version) +
        " hostname " + cleanString(hostname) +
        " pid " + cleanString(process.pid) +
        " runtime " + cleanString("node/" + process.versions.node) +
        " platform " + cleanString(process.platform + "-" + process.arch) + "\n" +
      "authenticate " + key + "\n"
    );
  });

  // We need to handle the timeout. I think we should only care about read
  // timeouts. That is we write out data, if we don't hear back in timeout
  // seconds then the data probably hasn't reached the server.
  client.setTimeout(timeout, function() {
    // ZOMG FAILED WRITING WE ARE BAD AT COMPUTER SCIENCE
    if(state == "connected") {
      // We're waiting to hear back from the server and it has timed out. It's
      // unlikely that the server will suddenly wake up and send us our data so
      // lets disconnect and go shopping.
      client.end();
    }
  });

  // HOW WE HANDLE ERRORS. We should probably reconnect, maybe retry.
  client.addListener("error", function(exception){
    if(debug) {
      log("Client error:", exception);
    }
    instrumentalStats.last_exception = Math.round(new Date().getTime() / 1000);
  });

  // What do we do when instrumental talks to us
  var totalBuffer = "";
  client.on("data", function(buffer) {
    totalBuffer = totalBuffer + buffer;

    if(debug) {
      log("Received:", buffer);
    }

    // Authorization success
    if(totalBuffer == "ok\nok\n") {
      error = false;

      if(debug) {
        log("Sending:", payload.join("\n"));
      }

      client.end(payload.join("\n") + "\n", function() {
        tlsOptionsVariation.lastSuccessAt = new Date();
        if (debug) log("payload sent, tlsOptionsVariation: ", tlsOptionsVariation);
        state = "sent";
      });

      instrumentalStats.last_flush = Math.round(new Date().getTime() / 1000);

    // Authorization failure
    } else if(totalBuffer.length >= "ok\nok\n".length) {
      // TODO: Actually do something with this
      error = true;

      instrumentalStats.last_exception = Math.round(new Date().getTime() / 1000);
    }
  });
}

function instrumental_flush(time_stamp, metrics) {
  if(flushInterval < 60000){
    metrics = buffering_aggregation(metrics, time_stamp, flushInterval);
  }
  var payload = build_payload(metrics, time_stamp);

  // Right now payload is at a minimum "\n"
  if(payload.length > 0) {
    instrumental_send(payload);
  }
}

function instrumental_status(writeCb) {
  for (stat in instrumentalStats) {
    writeCb(null, 'instrumental', stat, instrumentalStats[stat]);
  }
  writeCb(null, 'instrumental', 'error', error ? 1 : 0);
}

function knownValidCertNames() {
  var certNames = [];
  knownCerts.forEach(function(certName){
    var certExpiration = certName.split(".")[1];
    certExpiration = certExpiration && new Date(Date.parse(certExpiration));
    if ( ! certExpiration || (new Date() < certExpiration) ){
      certNames.push(certName);
    } else {
      if (debug) log("Ignoring expired cert bundle: " + certName);
    }
  });
  return certNames;
}

function knownValidCertPaths() {
  var certDir = path.join(__dirname, caChainPath);
  var caPaths = [];
  knownValidCertNames().forEach(function(certName){
    log("Found valid cert bundle: " + certName);
    var certPath = path.join(certDir, certName + ".ca.pem");
    caPaths.push(certPath);
  });
  return caPaths;
}

function certsFromPaths(paths) {
  var certs = [];
  paths.forEach(function(certPath){
    certs.push(fs.readFileSync(certPath));
  });
  return certs;
}

function initTlsOptionsVariations(config) {
  if (tlsOptionsVariations) { return tlsOptionsVariations; };

  var variations = [];

  if(typeof(config.instrumental.caCertFiles) === 'undefined'){
    // Node default certs
    if (config.instrumental.disallowNodeDefaultCerts) {
      log("Skipping node default certificates");
    } else {
      log("Adding node default ssl cert option");
      variations.push({ name: "node default", options: { rejectUnauthorized: verifyCert } });
    }

    // Cert bundles
    var certNames = knownValidCertNames();
    var certs = certsFromPaths(knownValidCertPaths());
    log("Adding known ssl bundles: " + certNames.join(" "));
    variations.push({ name: "bundles " + certNames.join(" "), options: { rejectUnauthorized: verifyCert, ca: certs } });
  } else {
    // User supplied, only use this
    config.instrumental.caCertFiles.forEach(function(certFile){
      log("Adding user supplied caCertFile option: " + certFile);
    });
    var certs = certsFromPaths(config.instrumental.caCertFiles);
    variations.push({ name: "user supplied", options: { rejectUnauthorized: verifyCert, ca: certs } });
  }

  tlsVariationTimeout = config.instrumental.tlsVariationTimeout || 10000;

  // Set the lastAttemptedAt to now so it's a comparable time. Set lastSuccessAt
  // in the order we would like these to be tried (used in sorting in
  // selectTlsOptionsVariation).
  variations.forEach(function(variation, index){
    // These should differ by more than tlsVariationTimeout so their initial
    // state is "failed" for the purposes of selectTlsOptionsVariation in order
    // to make the selection process more predictable.

    // A time in the order we want
    variation.lastAttemptedAt = new Date(index);

    // A time more than "failure" time before that
    variation.lastSuccessAt = new Date(index - tlsVariationTimeout*2);
  });

  tlsOptionsVariations = variations;
}

exports.init = function instrumental_init(startup_time, config, events) {
  debug = config.debug;

  if(config.instrumental) {
    if(typeof(config.instrumental.log) !== 'undefined'){
      log = config.instrumental.log;
    }

    key  = config.instrumental.key;
    host = config.instrumental.host || "collector.instrumentalapp.com";

    // Default 10s timeout
    timeout = Number(config.instrumental.timeout || 10000);

    // Record counter_rates by default
    if(typeof(config.instrumental.recordCounterRates) === 'undefined'){
      recordCounterRates = true;
    } else {
      recordCounterRates = config.instrumental.recordCounterRates;
    }

    // do no filtering by default
    if(typeof(config.instrumental.metricFiltersExclude) === 'undefined'){
      metricFiltersExclude = [];
    } else {
      metricFiltersExclude = config.instrumental.metricFiltersExclude;
    }

    // do no filtering by default
    if(typeof(config.instrumental.metricFiltersInclude) === 'undefined'){
      metricFiltersInclude = [];
    } else {
      metricFiltersInclude = config.instrumental.metricFiltersInclude;
    }
    keepMetric = function(metric){
      metricName = metric.split(" ")[1];
      return (!metricFiltersExclude.some(function(filter) { return metricName.match(filter); }) &&
              (metricFiltersInclude.length == 0 || metricFiltersInclude.some(function(filter) { return metricName.match(filter); })));
    };

    if(typeof(config.instrumental.metricPrefix) !== 'undefined' &&
       config.instrumental.metricPrefix != "" &&
       config.instrumental.metricPrefix != "."){
      var joinCharacter = ".";

      if(config.instrumental.metricPrefix.match(/\.$/))
      {
        joinCharacter = "";
      }

      formatMetricName = function(metric) {
        return [config.instrumental.metricPrefix, metric].join(joinCharacter);
      }
      metricPrefix = config.instrumental.metricPrefix;
    }

    if(typeof(config.instrumental.secure) === 'undefined'){
      secure = true;
    } else {
      secure = config.instrumental.secure;
    }

    // necessary while we transition config from verify_cert to verifyCert.
    if(typeof(config.instrumental.verifyCert) === 'undefined'){
      if(typeof(config.instrumental.verify_cert) === 'undefined'){
        verifyCert = true;
      } else {
        verifyCert = config.instrumental.verify_cert;
      }
    } else {
      verifyCert = config.instrumental.verifyCert;
    }
    if(secure && verifyCert){
      // Preserve previous setting for testing purposes
      initTlsOptionsVariations(config);
    }

    var defaultPort = secure ? 8001 : 8000;

    port = Number(config.instrumental.port || defaultPort);
  }

  flushInterval = config.flushInterval;

  // Check for credentials
  if(!key) {
    log("Missing instrumental.key from the config.");
    return false;
  }

  log("Connecting to " + host + ":" + port)

  instrumentalStats.last_flush = startup_time;
  instrumentalStats.last_exception = startup_time;

  events.on("flush", instrumental_flush);
  events.on("status", instrumental_status);
  dns.resolve(hostname, function(err, ips) {
    if (!err) {
      dns.reverse(ips[0], function (err, domains) {
        if (!err && String(domains[0]).length > 0) {
          hostname = domains[0];
        }
      });
    }
  });

  return true;
};
